### PR TITLE
Refactored GetCredentials Method

### PR DESCRIPTION
**Pull request by VibhorChinda**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Breaks down the  GetCredentials method into ValidateIdentity, CreateSecretKey, GetKey, GetNamespceKey

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by geetikabatra** [comment]
You might want to sign off this PR. 
Do and force push the change: `git commit --amend --signoff`

**Comment by VibhorChinda** [comment]
> You might want to sign off this PR. Do and force push the change: `git commit --amend --signoff`

@geetikabatra is this alright ?

**Comment by VibhorChinda** [comment]
@geetikabatra @aartij17 does failing test means the code is breaking somewhere and need to be looked upon ?
Need some feedback and direction on this please.
Thanks :v:

**Comment by geetikabatra** [comment]
@VibhorChinda Yes, you might want to run `make test` to check if the tests pass in your local.
Also, could you write unit tests for the changes you made above?

File: pkg/identity/identity.go [hunk 1]
```diff
@@ -40,30 +40,20 @@ type Credentials struct {
...
-	if ref == nil {
-		return nil, errors.New("IdentityRef is required")
+	if err := ValidateIdentity(ctx, c, cluster, controllerNamespace); err != nil {
+		return nil, err
 	}
 
+	ref := cluster.Spec.IdentityRef
 	secret := &apiv1.Secret{}
 	var secretKey client.ObjectKey
+
```
Comment (geetikabatra) [inline pkg/identity/identity.go:50]: nit: remove extra line

File: pkg/identity/identity.go [hunk 2]
```diff
@@ -112,6 +97,62 @@ func GetCredentials(ctx context.Context, c client.Client, cluster *infrav1.VSphe
...
+	if ref == nil {
+		return errors.New("IdentityRef is required")
+	}
+	if controllerNamespace == "" {
+		return errors.New("controllerNamespace is required")
+	}
+
+	return nil
+}
+
```
Comment (geetikabatra) [inline pkg/identity/identity.go:120]: You can add a unit test for this method in `identity_test.go`

File: pkg/identity/identity.go [hunk 3]
```diff
@@ -112,6 +97,62 @@ func GetCredentials(ctx context.Context, c client.Client, cluster *infrav1.VSphe
...
+		Name: ref.Name,
+	}
+}
+
+func GetNamespaceKey(cluster *infrav1.VSphereCluster) client.ObjectKey {
+	return client.ObjectKey{
+		Name: cluster.Namespace,
+	}
+}
+
```
Comment (geetikabatra) [inline pkg/identity/identity.go:155]: Please add tests for all the above methods.

**Review by geetikabatra — CHANGES_REQUESTED** [review]

**Comment by VibhorChinda** [comment]
> @VibhorChinda Yes, you might want to run `make test` to check if the tests pass in your local. Also, could you write unit tests for the changes you made above?

Ok will do that.
Yes, I will work on creating unit tests now. :v:

**Comment by VibhorChinda** [comment]
@srm09 will do all the changes being asked.
Just to be sure, only "ValidateInput" method should be created in this PR and test for that should be written right ?

I created all these static methods because the Issue description asked [https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/1441] to do so.

**Comment by VibhorChinda** [comment]
@geetikabatra @srm09 @aartij17 please check :)

**Comment by aartij17** [comment]
@VibhorChinda Can you take a look at the failed lint test. You can use `ctx` as the first argument to the function `validateInputs`

File: pkg/identity/identity_test.go [hunk 4]
```diff
@@ -184,20 +184,53 @@ var _ = Describe("GetCredentials", func() {
...
+				GenerateName: "cluster-",
+				Namespace:    ns.Name,
+			},
+		}
+		Expect(k8sclient.Create(ctx, cluster)).To(Succeed())
+	})
+
+	AfterEach(func() {
+		Expect(k8sclient.Delete(ctx, ns)).To(Succeed())
+		Expect(k8sclient.Delete(ctx, cluster)).To(Succeed())
```
Comment (srm09) [inline pkg/identity/identity_test.go:214]: Reverse the order, though I think, just deleting the namespace should delete the objects contained in the namespace automaGically.

File: pkg/identity/identity_test.go [hunk 5]
```diff
@@ -184,20 +184,53 @@ var _ = Describe("GetCredentials", func() {
...
+	})
+
+	Context("If the client is missing", func() {
+		It("should error if client is missing", func() {
+			err := validateInputs(nil, cluster)
 			Expect(err).To(HaveOccurred())
```
Comment (srm09) [inline pkg/identity/identity_test.go:220]:
```suggestion
			Expect(validateInputs(nil, cluster)).NotTo(Succeed())
```

File: pkg/identity/identity_test.go [hunk 6]
```diff
@@ -184,20 +184,53 @@ var _ = Describe("GetCredentials", func() {
...
 		})
+	})
 
-		It("should error if client is missing", func() {
-			_, err := GetCredentials(ctx, nil, cluster, manager.DefaultPodNamespace)
+	Context("If the cluster is missing", func() {
+		It("should error if cluster is missing", func() {
+			err := validateInputs(k8sclient, nil)
```
Comment (srm09) [inline pkg/identity/identity_test.go:226]: Same as above

File: pkg/identity/identity_test.go [hunk 7]
```diff
@@ -184,20 +184,53 @@ var _ = Describe("GetCredentials", func() {
...
 			Expect(err).To(HaveOccurred())
 		})
+	})
 
+	Context("If the identityRef is missing on cluster", func() {
 		It("should error if identityRef is missing on cluster", func() {
-			_, err := GetCredentials(ctx, k8sclient, cluster, manager.DefaultPodNamespace)
+			err := validateInputs(k8sclient, cluster)
```
Comment (srm09) [inline pkg/identity/identity_test.go:233]: Same as above

**Review by yastij — APPROVED** [review]
/approve

**Comment by VibhorChinda** [comment]
First CAPV PR merged 🎉🎉

**Pull request merged**